src/main/java/uk/gov/register/serialization/CommandParser.java

```diff
@@ -45,40 +59,84 @@ public RegisterCommand newCommand(String s){
                         Item item = new Item(itemHash, objectReconstructor.reconstruct(jsonContent));
-                        return new AddItemCommand(item);
-                    } catch (JsonParseException jpe){
+                        items.put(position++, item);
+                        itemHashToEntryCount.put(itemHash, 0);
+                    } catch (JsonParseException jpe) {
                         LOG.error("failed to parse json: " + parts[1]);
                         throw new SerializedRegisterParseException("failed to parse json: " + parts[1], jpe);
                     } catch (IOException e) {
-                        LOG.error("",e);
+                        LOG.error("", e);
                         throw new UncheckedIOException(e);
                     }
                 } else {
                     LOG.error("add item line must have 2 elements, was: " + s);
                     throw new SerializedRegisterParseException("add item line must have 2 elements, was: " + s);
                 }
+                break;
             case "append-entry":
                 if (parts.length == 3) {
                     Entry entry = new Entry(0, stripPrefix(parts[2]), Instant.parse(parts[1]));
-                    return new AppendEntryCommand(entry);
+                    entries.put(position++, entry);
+                    updateItemHashCount(entry.getSha256hex());
                 } else {
                     LOG.error("append entry line must have 3 elements, was : " + s);
                     throw new SerializedRegisterParseException("append entry line must have 3 elements, was : " + s);
                 }
+                break;
             case "assert-root-hash":
                 if (parts.length == 2) {
-                    return new AssertRootHashCommand(new RegisterProof(parts[1]));
+                    RegisterProof registerProof = new RegisterProof(parts[1]);
+                    proofs.put(position++, registerProof);
                 } else {
                     LOG.error("assert root hash line must have 1 elements, was : " + s);
                     throw new SerializedRegisterParseException("assert root hash line must have 1 elements, was : " + s);
                 }
+                break;
             default:
                 LOG.error("line must begin with legal command not:" + commandName);
                 throw new SerializedRegisterParseException("line must begin with legal command not: " + commandName);
         }
     }
 
-    private  String stripPrefix(String hashField) {
+    private void updateItemHashCount(String sha256hex) {
+        if (itemHashToEntryCount.containsKey(sha256hex)) {
+            Integer count = itemHashToEntryCount.get(sha256hex);
+            itemHashToEntryCount.replace(sha256hex, count + 1);
+        }
+    }
+
+    public Iterator<RegisterCommand> getCommands() {
+        validateOrphanItems();
+
+        return IntStream.range(0, position).mapToObj(i -> {
```

Comment:
If we later find out that `AssertRootHash` is only going to appear at the start and end of an RSF file, we could revert to adding the set of items to the register first, after which we then go through the (ordered) list and add entries. Happy to merge this for the time being though.